airbyte-cdk/python/airbyte_cdk/test/mock_http/mocker.py

```diff
@@ -82,7 +89,13 @@ def wrapper(*args, **kwargs):  # type: ignore  # this is a very generic wrapper
 
                 # We validate the matchers before raising the assertion error because we want to show the tester if a HTTP request wasn't
                 # mocked correctly
-                self._validate_all_matchers_called()
+                try:
+                    self._validate_all_matchers_called()
+                except ValueError as http_mocker_exception:
+                    # This seems useless as it catches ValueError and raises ValueError but without this, the prevaling error message in
```

Comment:
Timebox: This is awkward because I can't explain exactly why and I don't know how to test that. I think for not it's good enough to comment to explain why it's there but if this is a recurrent problem in the future, we should invest more time to understand why and add some tests

Reply:
I don't understand what this means. I tried removing the try/except and all the tests still pass. what's the difference in outcome you're seeing?

Reply:
Let's say I'm a developer and I set up a test where there is an error in the HTTP request:
```
    @HttpMocker()
    def test_given_one_page_when_read_then_return_record(self, http_mocker: HttpMocker) -> None:
        http_mocker.get(
            HttpRequest(
                url="https://api.stripe.com/v1/event",
                query_params={"created[gte]": str(int(_A_START_DATE.timestamp())), "created[lte]": str(int(_NOW.timestamp())), "limit": 100},
                headers=_AUTHENTICATION_HEADERS
            ),
            _a_response().with_record(_a_record()).build()
        )
        output = self._read(_config().with_start_date(_A_START_DATE))
        assert len(output.records) == 1
```
In this case, I changed `https://api.stripe.com/v1/events` to `https://api.stripe.com/v1/event`. Without this weird try/exception the error will be:

```
Printing unexpected error from entrypoint_wrapper
Traceback (most recent call last):
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/test/entrypoint_wrapper.py", line 145, in read
    for message in source_entrypoint.run(parsed_args):
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py", line 120, in run
    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py", line 162, in read
    yield from self.source.read(self.logger, config, catalog, state)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py", line 149, in read
    raise e
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py", line 120, in read
    stream_is_available, reason = stream_instance.check_availability(logger, self)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py", line 222, in check_availability
    availability = self._abstract_stream.check_availability()
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/default_stream.py", line 45, in check_availability
    return self._availability_strategy.check_availability(self._logger)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py", line 426, in check_availability
    raise e
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py", line 416, in check_availability
    available, message = self._stream.check_availability(logger, self._source)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py", line 263, in check_availability
    return self.availability_strategy.check_availability(self, logger, source)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/availability_strategy.py", line 75, in check_availability
    is_available, reason = self._check_availability_for_sync_mode(stream, SyncMode.full_refresh, logger, source, None)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/availability_strategy.py", line 51, in _check_availability_for_sync_mode
    get_first_record_for_slice(stream, sync_mode, stream_slice, stream_state)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/stream_helpers.py", line 41, in get_first_record_for_slice
    return next(records_for_slice)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 482, in read_records
    yield from self._read_pages(
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 498, in _read_pages
    request, response = self._fetch_next_page(stream_slice, stream_state, next_page_token)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 524, in _fetch_next_page
    response = self._send_request(request, request_kwargs)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 422, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 358, in _send
    response: requests.Response = self._session.send(request, **request_kwargs)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py", line 513, in send
    response = super().send(request, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/requests_mock/mocker.py", line 144, in _fake_send
    return _original_send(session, request, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/requests_mock/adapter.py", line 261, in send
    raise exceptions.NoMockAddress(request)
requests_mock.exceptions.NoMockAddress: No mock address: GET https://api.stripe.com/v1/events?created%5Bgte%5D=1697203465&created%5Blte%5D=1702387465&limit=100

FAILED
airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:80 (FullRefreshTest.test_given_one_page_when_read_then_return_record)
0 != 1

Expected :1
Actual   :0
<Click to see difference>

../../../../../airbyte-cdk/python/airbyte_cdk/test/mock_http/mocker.py:92: in wrapper
    self._validate_all_matchers_called()
```

As you can see in the end, we only have a very simple message mentioning `self._validate_all_matchers_called()` but it does not say why it's mentioned. With this weird try/exception, the error will be:
```
Printing unexpected error from entrypoint_wrapper
Traceback (most recent call last):
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/test/entrypoint_wrapper.py", line 145, in read
    for message in source_entrypoint.run(parsed_args):
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py", line 120, in run
    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/entrypoint.py", line 162, in read
    yield from self.source.read(self.logger, config, catalog, state)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py", line 149, in read
    raise e
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py", line 120, in read
    stream_is_available, reason = stream_instance.check_availability(logger, self)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py", line 222, in check_availability
    availability = self._abstract_stream.check_availability()
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/default_stream.py", line 45, in check_availability
    return self._availability_strategy.check_availability(self._logger)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py", line 426, in check_availability
    raise e
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py", line 416, in check_availability
    available, message = self._stream.check_availability(logger, self._source)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py", line 263, in check_availability
    return self.availability_strategy.check_availability(self, logger, source)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/availability_strategy.py", line 75, in check_availability
    is_available, reason = self._check_availability_for_sync_mode(stream, SyncMode.full_refresh, logger, source, None)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/availability_strategy.py", line 51, in _check_availability_for_sync_mode
    get_first_record_for_slice(stream, sync_mode, stream_slice, stream_state)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/stream_helpers.py", line 41, in get_first_record_for_slice
    return next(records_for_slice)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 482, in read_records
    yield from self._read_pages(
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 498, in _read_pages
    request, response = self._fetch_next_page(stream_slice, stream_state, next_page_token)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 524, in _fetch_next_page
    response = self._send_request(request, request_kwargs)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 422, in _send_request
    return backoff_handler(user_backoff_handler)(request, request_kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/backoff/_sync.py", line 105, in retry
    ret = target(*args, **kwargs)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py", line 358, in _send
    response: requests.Response = self._session.send(request, **request_kwargs)
  File "./airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py", line 513, in send
    response = super().send(request, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/requests_mock/mocker.py", line 144, in _fake_send
    return _original_send(session, request, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "./airbyte/airbyte-integrations/connectors/source-stripe/.venv/lib/python3.9/site-packages/requests_mock/adapter.py", line 261, in send
    raise exceptions.NoMockAddress(request)
requests_mock.exceptions.NoMockAddress: No mock address: GET https://api.stripe.com/v1/events?created%5Bgte%5D=1697203738&created%5Blte%5D=1702387738&limit=100

FAILED
airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py:80 (FullRefreshTest.test_given_one_page_when_read_then_return_record)
0 != 1

Expected :1
Actual   :0
<Click to see difference>

args = (<integration.test_events.FullRefreshTest testMethod=test_given_one_page_when_read_then_return_record>,)
kwargs = {'http_mocker': <airbyte_cdk.test.mock_http.mocker.HttpMocker object at 0x110af6520>}
assertion_error = AssertionError('assert 0 == 1\n  +0\n  -1')

    @functools.wraps(f)
    def wrapper(*args, **kwargs):  # type: ignore  # this is a very generic wrapper that does not need to be typed
        with self:
            assertion_error = None
    
            kwargs["http_mocker"] = self
            try:
                result = f(*args, **kwargs)
            except requests_mock.NoMockAddress as no_mock_exception:
                matchers_as_string = "\n\t".join(map(lambda matcher: str(matcher.request), self._matchers))
                raise ValueError(
                    f"No matcher matches {no_mock_exception.args[0]} with headers `{no_mock_exception.request.headers}`. Matchers currently configured are:\n\t{matchers_as_string}"
                ) from no_mock_exception
            except AssertionError as test_assertion:
                assertion_error = test_assertion
    
            # We validate the matchers before raising the assertion error because we want to show the tester if a HTTP request wasn't
            # mocked correctly
            try:
                self._validate_all_matchers_called()
            except ValueError as http_mocker_exception:
                # This seems useless as it catches ValueError and raises ValueError but without this, the prevaling error message in
                # the output is the assertion. The error message for the matchers validation is only
                # `../../../../../airbyte-cdk/python/airbyte_cdk/test/mock_http/mocker.py:92: in wrapper\n\tself._validate_all_matchers_called()`
>               raise ValueError(http_mocker_exception) from None
E               ValueError: Invalid number of matches for `HttpRequestMatcher(request_to_match=ParseResult(scheme='https', netloc='api.stripe.com', path='/v1/event', params='', query='created%5Bgte%5D=1697203738&created%5Blte%5D=1702387738&limit=100', fragment='') with headers {'Stripe-Account': 'account_id', 'Authorization': 'Bearer client_secret'}), minimum_number_of_expected_match=1, actual_number_of_matches=0)`

../../../../../airbyte-cdk/python/airbyte_cdk/test/mock_http/mocker.py:98: ValueError
```

... where the actual error is printed and the message is explicit. Again, I've tried to understand why such behavior occur but couldn't so I simply timeboxed. Let me know if you think it's worth investigating more or how I can improve the comment

Reply:
I took a stab at rewording it in a comment below.

Reply:
The rewording is indeed better. As long as @girarda confirms that it's enough for him to explain the situation, I'll resolve this comment

Reply:
lgtm thanks Catherine and Maxime